core/src/forwarding_stage.rs

```diff
@@ -38,73 +49,267 @@ mod packet_container;
-    /// Forwarding address for votes
-    pub tpu_vote: Option<SocketAddr>,
-}
-
-pub trait ForwardAddressGetter: Send + Sync + 'static {
-    fn get_forwarding_addresses(&self, protocol: Protocol) -> ForwardingAddresses;
+// One function has been split into two so that VoteClient and NonVoteClient
+// could manage their addresses internally, providing better encapsulation.
+pub trait ForwardAddressGetter: Clone + Send + Sync + 'static {
+    // Why added `max_count` -- because tpu-client-next refreshes connections
+    // before using them by trying to connect, while ConnectionCache refreshes
+    // them separately in a different thread by sending empty packets.
+    fn get_non_vote_forwarding_addresses(
+        &self,
+        max_count: u64,
+        protocol: Protocol,
+    ) -> Vec<Option<SocketAddr>>;
+    fn get_vote_forwarding_addresses(&self) -> Option<SocketAddr>;
 }
 
 impl<T: LikeClusterInfo> ForwardAddressGetter for (T, Arc<RwLock<PohRecorder>>) {
-    fn get_forwarding_addresses(&self, protocol: Protocol) -> ForwardingAddresses {
-        ForwardingAddresses {
-            tpu: next_leader(&self.0, &self.1, |node| node.tpu_forwards(protocol)).map(|(_, s)| s),
-            tpu_vote: next_leader_tpu_vote(&self.0, &self.1).map(|(_, s)| s),
-        }
+    fn get_non_vote_forwarding_addresses(
+        &self,
+        max_count: u64,
+        protocol: Protocol,
+    ) -> Vec<Option<SocketAddr>> {
+        next_leaders(&self.0, &self.1, max_count, |node| {
+            node.tpu_forwards(protocol)
+        })
+    }
+    fn get_vote_forwarding_addresses(&self) -> Option<SocketAddr> {
+        next_leader_tpu_vote(&self.0, &self.1).map(|(_, s)| s)
     }
 }
 
-struct VoteClient {
+pub trait ForwardingClient: Send + Sync + 'static {
+    fn update_address(&mut self) -> bool;
+    fn send_batch(&self, input_batch: Vec<Vec<u8>>);
+}
+
+struct VoteClient<F: ForwardAddressGetter> {
     udp_socket: UdpSocket,
+    forward_address_getter: F,
+    current_address: Option<SocketAddr>,
 }
 
-impl VoteClient {
-    fn new() -> Self {
+impl<F: ForwardAddressGetter> VoteClient<F> {
+    fn new(forward_address_getter: F) -> Self {
         Self {
             udp_socket: bind_to_unspecified().unwrap(),
+            forward_address_getter,
+            current_address: None,
         }
     }
 
-    fn send_batch(&self, batch: &mut Vec<(Vec<u8>, SocketAddr)>) {
-        let pkts = batch.iter().map(|(bytes, addr)| (bytes, addr));
-        let _res = batch_send(&self.udp_socket, pkts);
+    fn update_address(&mut self) -> bool {
+        self.current_address = self.forward_address_getter.get_vote_forwarding_addresses();
+        self.current_address.is_some()
+    }
+
+    // Why it doesn't make sense to pass pairs of (Vec<u8>, SocketAddr):
+    // because we actually always send to the same address, this generic argument making
+    // a false impression.
+    // Why not implementing the same ForwardingClient trait -- to have batch: &mut
```

Comment:
I don't think we need a comment here - it's just fixing a bad previous interface I moved from forwarder